src/shims/foreign_items.rs

```diff
@@ -507,6 +507,70 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 this.write_scalar(Scalar::from_int(result, dest.layout.size), dest)?;
             }
 
+            "clock_gettime" => {
+                if !this.machine.communicate {
+                    throw_unsup_format!("`clock_gettime` not available when isolation is enabled")
+                } else {
+                    let clk_id = this.read_scalar(args[0])?.to_i32()?;
+
+                    if clk_id != this.eval_libc_i32("CLOCK_REALTIME")? {
+                        let einval = this.eval_libc("EINVAL")?;
+                        this.set_last_error(einval)?;
+                        this.write_scalar(Scalar::from_int(-1i32, dest.layout.size), dest)?;
+                    } else {
+                        let tp = this.force_ptr(this.read_scalar(args[1])?.not_undef()?)?;
+
+                        let mut sign = 1;
+
+                        let duration = std::time::SystemTime::now()
+                            .duration_since(std::time::SystemTime::UNIX_EPOCH)
+                            .unwrap_or_else(|e| {
+                                sign = -1;
+                                e.duration()
+                            });
+
+                        let tv_sec = sign * (duration.as_secs() as i128);
+                        let tv_nsec = duration.subsec_nanos() as i128;
+
+                        this.write_c_ints(&tp, &[tv_sec, tv_nsec], &["time_t", "c_long"])?;
+
+                        this.write_scalar(Scalar::from_int(0i32, dest.layout.size), dest)?;
+                    }
+                }
```

Comment:
`foreign_items` should only contain *trivial* implementations of things. So when stuff gets more complicated, like this, please move them to a separate file. In this case a new file like `clock.rs` seems to be the best choice.

(I did not look at your file system PRs yet, but if they are implemented inline here, the same concern applies.)

Reply:
Ok I'll take this out to its own file. All the logic for the file related stuff is in its own module inside shims too.